presto-spark/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskCompiler.java

```diff
@@ -0,0 +1,340 @@
+            ScheduledExecutorService yieldExecutor,
+            LocalExecutionPlanner localExecutionPlanner,
+            BlockEncodingSerde blockEncodingSerde,
+            TaskManagerConfig taskManagerConfig,
+            NodeMemoryConfig nodeMemoryConfig,
+            NodeSpillConfig nodeSpillConfig)
+    {
+        this(
+                sessionPropertyManager,
+                sparkTaskDescriptorJsonCodec,
+                taskStatsJsonCodec,
+                notificationExecutor,
+                yieldExecutor,
+                localExecutionPlanner,
+                blockEncodingSerde,
+                requireNonNull(nodeMemoryConfig, "nodeMemoryConfig is null").getMaxQueryMemoryPerNode(),
+                requireNonNull(nodeMemoryConfig, "nodeMemoryConfig is null").getMaxQueryTotalMemoryPerNode(),
+                requireNonNull(nodeSpillConfig, "nodeSpillConfig is null").getMaxSpillPerNode(),
+                requireNonNull(taskManagerConfig, "taskManagerConfig is null").isPerOperatorCpuTimerEnabled(),
+                requireNonNull(taskManagerConfig, "taskManagerConfig is null").isTaskCpuTimerEnabled());
+    }
+
+    public PrestoSparkTaskCompiler(
+            SessionPropertyManager sessionPropertyManager,
+            JsonCodec<SparkTaskDescriptor> sparkTaskDescriptorJsonCodec,
+            JsonCodec<TaskStats> taskStatsJsonCodec,
+            Executor notificationExecutor,
+            ScheduledExecutorService yieldExecutor,
+            LocalExecutionPlanner localExecutionPlanner,
+            BlockEncodingSerde blockEncodingSerde,
+            DataSize maxUserMemory,
+            DataSize maxTotalMemory,
+            DataSize maxSpillMemory,
+            boolean perOperatorCpuTimerEnabled,
+            boolean cpuTimerEnabled)
+    {
+        this.sessionPropertyManager = requireNonNull(sessionPropertyManager, "sessionPropertyManager is null");
+        this.sparkTaskDescriptorJsonCodec = requireNonNull(sparkTaskDescriptorJsonCodec, "sparkTaskDescriptorJsonCodec is null");
+        this.taskStatsJsonCodec = requireNonNull(taskStatsJsonCodec, "taskStatsJsonCodec is null");
+        this.notificationExecutor = requireNonNull(notificationExecutor, "notificationExecutor is null");
+        this.yieldExecutor = requireNonNull(yieldExecutor, "yieldExecutor is null");
+        this.localExecutionPlanner = requireNonNull(localExecutionPlanner, "localExecutionPlanner is null");
+        this.blockEncodingSerde = requireNonNull(blockEncodingSerde, "blockEncodingSerde is null");
+        this.maxUserMemory = requireNonNull(maxUserMemory, "maxUserMemory is null");
+        this.maxTotalMemory = requireNonNull(maxTotalMemory, "maxTotalMemory is null");
+        this.maxSpillMemory = requireNonNull(maxSpillMemory, "maxSpillMemory is null");
+        this.perOperatorCpuTimerEnabled = perOperatorCpuTimerEnabled;
+        this.cpuTimerEnabled = cpuTimerEnabled;
+    }
+
+    @Override
+    public Iterator<Tuple2<Integer, byte[]>> compile(
```

Comment:
`createSqlExecutor`?

Reply:
Note this `Iterator<Tuple2<Integer, byte[]>>` is not easy to understand, since it's a bit far away from the actual `flatMapToPair` call.

One idea is to create a dumb interface `ISparkSqlExecutor` that extends the iterator: 
```java
public interface ISparkSqlExecutor extends Iterator<Tuple2<Integer, byte[]>
```

And make this returns a `ISparkSqlExecutor`.  For `SparkSqlExecutor`, it will be like this:
```java
    public static class SparkSqlExecutor
             extends AbstractIterator<Tuple2<Integer, byte[]>>
            implements ISparkSqlExecutor
```

So basically I am using `ISparkSqlExecutor` as a "typedef" in C++ :) .

Here is a diff: https://gist.github.com/wenleix/ffd91b45e41bdce2e95aa2f36c26ea5d seems to work :)

Reply:
>createSqlExecutor

I renamed the method to simply `create`

Reply:
For the second - i created the `SparkExecutor` interface in the `presto-spark-classloader-interface`. I don't think we need the `ISparkExecutor`, as the `SparkExecutor` is already an interface.